src/aosm/azext_aosm/generate_nfd/cnf_nfd_generator.py

```diff
@@ -15,7 +37,374 @@ class CnfNfdGenerator(NFDGenerator):
+
+    def find_pattern_matches_in_chart(
+        self, helm_package: HelmPackageConfig, pattern: str
+    ) -> List[Tuple[str, ...]]:
+        chart_dir = os.path.join(self.tmp_folder_name, helm_package.name)
+        matches = []
+
+        for file in self._find_yaml_files(chart_dir):
+            with open(file, "r", encoding="UTF-8") as f:
+                contents = f.read()
+                matches += re.findall(pattern, contents)
+
+        return matches
+
+    def get_artifact_list(
+        self,
+        helm_package: HelmPackageConfig,
+        image_line_matches: List[Tuple[str, ...]],
+    ) -> List[Any]:
+        artifact_list = []
+        (chart_name, chart_version) = self.get_chart_name_and_version(
+            helm_package
+        )
+        helm_artifact = {
+            "name": chart_name,
+            "version": chart_version,
+        }
+        artifact_list.append(helm_artifact)
+
+        for match in image_line_matches:
+            artifact_list.append(
+                {
+                    "name": match[1],
+                    "version": match[2],
+                }
+            )
+
+        return artifact_list
+
+    def get_chart_mapping_schema(
+        self, helm_package: HelmPackageConfig
+    ) -> Dict[Any, Any]:
+        non_def_values = os.path.join(
+            self.tmp_folder_name, helm_package.name, "values.mappings.yaml"
+        )
+        values_schema = os.path.join(
+            self.tmp_folder_name, helm_package.name, "values.schema.json"
+        )
+
+        with open(non_def_values, "r", encoding="utf-8") as stream:
+            values_data = yaml.load(stream, Loader=yaml.SafeLoader)
+
+        with open(values_schema, "r", encoding="utf-8") as f:
+            data = json.load(f)
+            schema_data = data["properties"]
+
+        try:
+            final_schema = self.find_deploy_params(
+                values_data, schema_data, {}
+            )
+        except KeyError as e:
+            raise InvalidTemplateError(
+                f"ERROR: Your schema and values for the helm package '{helm_package.name}' do not match. Please fix this and run the command again."
+            ) from e
+
+        return final_schema
+
+    def find_deploy_params(
+        self, nested_dict, schema_nested_dict, final_schema
+    ) -> Dict[Any, Any]:
+        original_schema_nested_dict = schema_nested_dict
+        for k, v in nested_dict.items():
+            # if value is a string and contains deployParameters.
+            if isinstance(v, str) and re.search(DEPLOYMENT_PARAMETER_MAPPING_REGEX, v):
+                
+                # only add the parameter name (e.g. from {deployParameter.zone} only param = zone)
+                param = v.split(".", 1)[1]
+                param = param.split("}", 1)[0]
+
+                # add the schema for k (from the big schema) to the (smaller) schema
+                final_schema.update({param: { "type" : schema_nested_dict["properties"][k]["type"]}})
```

Comment:
I'd like you to get a second opinion from another python expert (Chaos?) about whether this is a good thing. We are passing in a dictionary to the function which both gets updated in the function and returned from it.  That feels a little odd, but might be fine, I'm not sure.  Passing in a thing which gets updated is often a recipe for errors and should definitely be highlighted in any docstring for this function if we do leave it as it is.

Reply:
Will implement Chaos's changes when i get a chance, but for now will merge